File: src/main/java/dev/lain/claudejb/mcp/ToonEncoder.java
package dev.lain.claudejb.mcp;

import dev.lain.claudejb.mcp.ToonHeader.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

final class ToonEncoder {

    private final Toon.Options options;
    private final char delimiter;
    private final String symbol;
    private final List<String> lines = new ArrayList<>();

    ToonEncoder(Toon.Options options) {
        this.options = options;
        this.delimiter = options.delimiter;
        this.symbol = delimiter == ',' ? "" : String.valueOf(delimiter);
    }

    String encode(Object value) {
        if (value instanceof List) array("", (List<?>) value, 0);
        else if (value instanceof Map) {
            Map<String, ?> obj = cast(value);
            List<Field> columns = keyedColumns(obj);
            if (columns != null) keyed("", obj, columns, 0);
            else fields(obj, 0);
        } else lines.add(ToonText.primitive(value, delimiter));
        return String.join("\n", lines);
    }

    private void fields(Map<String, ?> obj, int depth) {
        for (Map.Entry<String, ?> entry : obj.entrySet()) field(ToonText.key(entry.getKey()), entry.getValue(), depth);
    }

    private void field(String name, Object value, int depth) {
        if (value instanceof List) array(name, (List<?>) value, depth);
        else if (value instanceof Map) object(name, cast(value), depth);
        else emit(depth, name + ": " + ToonText.primitive(value, delimiter));
    }

    private void object(String name, Map<String, ?> value, int depth) {
        List<Field> columns = keyedColumns(value);
        if (value.isEmpty()) emit(depth, name + ":");
        else if (columns != null) keyed(name, value, columns, depth);
        else {
            emit(depth, name + ":");
            fields(value, depth + 1);
        }
    }

    private void array(String name, List<?> arr, int depth) {
        List<Field> columns = !arr.isEmpty() && allObjects(arr) ? columns(arr) : null;
        if (arr.isEmpty()) emit(depth, name.isEmpty() ? "[]" : name + ": []");
        else if (allPrimitives(arr)) emit(depth, name + bracket(arr.size()) + ": " + inline(arr));
        else if (columns != null) {
            emit(depth, name + bracket(arr.size()) + "{" + ToonHeader.render(columns, delimiter) + "}:");
            for (Object row : arr) emit(depth + 1, String.join(String.valueOf(delimiter), cells(cast(row), columns)));
        } else {
            emit(depth, name + bracket(arr.size()) + ":");
            for (Object element : arr) item(element, depth + 1);
        }
    }

    private void keyed(String name, Map<String, ?> obj, List<Field> columns, int depth) {
        emit(depth, name + "[" + obj.size() + ":" + symbol + "]{" + ToonHeader.render(columns, delimiter) + "}:");
        for (Map.Entry<String, ?> entry : obj.entrySet()) {
            String cells = String.join(String.valueOf(delimiter), cells(cast(entry.getValue()), columns));
            emit(depth + 1, ToonText.key(entry.getKey()) + ": " + cells);
        }
    }

    private void item(Object element, int depth) {
        if (element instanceof List) arrayItem((List<?>) element, depth);
        else if (element instanceof Map) objectItem(cast(element), depth);
        else emit(depth, "- " + ToonText.primitive(element, delimiter));
    }

    private void arrayItem(List<?> element, int depth) {
        if (allPrimitives(element)) {
            String values = element.isEmpty() ? "" : " " + inline(element);
            emit(depth, "- " + bracket(element.size()) + ":" + values);
        } else {
            emit(depth, "- " + bracket(element.size()) + ":");
            for (Object nested : element) item(nested, depth + 1);
        }
    }

    private void objectItem(Map<String, ?> element, int depth) {
        if (element.isEmpty()) {
            emit(depth, "-");
            return;
        }
        int first = lines.size();
        fields(element, depth + 1);
        lines.set(first, indent(depth) + "- " + lines.get(first).substring(indent(depth + 1).length()));
    }

    private List<Field> columns(List<?> objects) {
        Set<String> keys = cast(objects.get(0)).keySet();
        if (keys.isEmpty()) return null;
        for (Object object : objects) if (!cast(object).keySet().equals(keys)) return null;
        List<Field> out = new ArrayList<>();
        for (String key : keys) {
            List<Object> values = new ArrayList<>();
            for (Object object : objects) values.add(cast(object).get(key));
            Field column = column(key, values);
            if (column == null) return null;
            out.add(column);
        }
        return out;
    }

    private Field column(String name, List<Object> values) {
        if (allPrimitives(values)) return new Field(name, null);
        if (!allObjects(values)) return null;
        List<Field> children = columns(values);
        return children == null ? null : new Field(name, children);
    }

    private List<Field> keyedColumns(Map<String, ?> obj) {
        if (obj.size() < 2 || !allObjects(obj.values())) return null;
        return columns(new ArrayList<>(obj.values()));
    }

    private List<String> cells(Map<String, ?> obj, List<Field> columns) {
        List<String> out = new ArrayList<>();
        for (Field column : columns) {
            Object value = obj.get(column.name);
            if (column.children == null) out.add(ToonText.primitive(value, delimiter));
            else out.addAll(cells(cast(value), column.children));
        }
        return out;
    }

    private String inline(List<?> arr) {
        List<String> out = new ArrayList<>();
        for (Object value : arr) out.add(ToonText.primitive(value, delimiter));
        return String.join(String.valueOf(delimiter), out);
    }

    private static boolean allPrimitives(Iterable<?> values) {
        for (Object value : values) if (!ToonText.isPrimitive(value)) return false;
        return true;
    }

    private static boolean allObjects(Iterable<?> values) {
        for (Object value : values) if (!(value instanceof Map)) return false;
        return true;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, ?> cast(Object value) {
        return (Map<String, ?>) value;
    }

    private String bracket(int size) {
        return "[" + size + symbol + "]";
    }

    private void emit(int depth, String text) {
        lines.add(indent(depth) + text);
    }

    private String indent(int depth) {
        return " ".repeat(depth * options.indentSize);
    }
}

File: src/main/kotlin/dev/lain/claudejb/model/mcp/toon/ToonEncoder.kt
package dev.lain.claudejb.model.mcp.toon

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

internal class ToonEncoder(private val options: ToonOptions) {

    private val delimiter = options.delimiter.symbol
    private val symbol = if (options.delimiter == ToonDelimiter.COMMA) "" else delimiter.toString()
    private val lines = ArrayList<String>()

    fun encode(value: JsonElement): String {
        when (value) {
            is JsonPrimitive -> lines += ToonText.primitive(value, delimiter)
            is JsonArray -> array("", value, 0)
            is JsonObject -> keyedColumns(value)?.let { keyed("", value, it, 0) } ?: fields(value, 0)
        }
        return lines.joinToString("\n")
    }

    private fun fields(obj: JsonObject, depth: Int) {
        for ((key, value) in obj) field(ToonText.key(key), value, depth)
    }

    private fun field(name: String, value: JsonElement, depth: Int) {
        when (value) {
            is JsonPrimitive -> emit(depth, "$name: ${ToonText.primitive(value, delimiter)}")
            is JsonArray -> array(name, value, depth)
            is JsonObject -> obj(name, value, depth)
        }
    }

    private fun obj(name: String, value: JsonObject, depth: Int) {
        val columns = keyedColumns(value)
        when {
            value.isEmpty() -> emit(depth, "$name:")

            columns != null -> keyed(name, value, columns, depth)

            else -> {
                emit(depth, "$name:")
                fields(value, depth + 1)
            }
        }
    }

    private fun array(name: String, arr: JsonArray, depth: Int) {
        val columns = if (arr.isNotEmpty() && arr.all { it is JsonObject }) columns(arr.map { it as JsonObject }) else null
        when {
            arr.isEmpty() -> emit(depth, if (name.isEmpty()) "[]" else "$name: []")

            arr.all { it is JsonPrimitive } -> emit(depth, "$name${bracket(arr.size)}: ${inline(arr)}")

            columns != null -> {
                emit(depth, "$name${bracket(arr.size)}{${ToonHeader.render(columns, delimiter)}}:")
                for (row in arr) emit(depth + 1, cells(row as JsonObject, columns).joinToString(delimiter.toString()))
            }

            else -> {
                emit(depth, "$name${bracket(arr.size)}:")
                for (element in arr) item(element, depth + 1)
            }
        }
    }

    private fun keyed(name: String, obj: JsonObject, columns: List<Field>, depth: Int) {
        emit(depth, "$name[${obj.size}:$symbol]{${ToonHeader.render(columns, delimiter)}}:")
        for ((key, value) in obj) {
            emit(depth + 1, "${ToonText.key(key)}: ${cells(value as JsonObject, columns).joinToString(delimiter.toString())}")
        }
    }

    private fun item(element: JsonElement, depth: Int) {
        when (element) {
            is JsonPrimitive -> emit(depth, "- ${ToonText.primitive(element, delimiter)}")
            is JsonArray -> arrayItem(element, depth)
            is JsonObject -> objectItem(element, depth)
        }
    }

    private fun arrayItem(element: JsonArray, depth: Int) {
        when {
            element.all { it is JsonPrimitive } -> {
                val values = if (element.isEmpty()) "" else " ${inline(element)}"
                emit(depth, "- ${bracket(element.size)}:$values")
            }

            else -> {
                emit(depth, "- ${bracket(element.size)}:")
                for (nested in element) item(nested, depth + 1)
            }
        }
    }

    private fun objectItem(element: JsonObject, depth: Int) {
        if (element.isEmpty()) {
            emit(depth, "-")
            return
        }
        val first = lines.size
        fields(element, depth + 1)
        lines[first] = indent(depth) + "- " + lines[first].substring(indent(depth + 1).length)
    }

    private fun columns(objects: List<JsonObject>): List<Field>? {
        val keys = objects.first().keys
        if (keys.isEmpty() || objects.any { it.keys != keys }) return null
        return keys.map { key -> column(key, objects.map { it.getValue(key) }) ?: return null }
    }

    private fun column(name: String, values: List<JsonElement>): Field? = when {
        values.all { it is JsonPrimitive } -> Field(name, null)
        values.all { it is JsonObject } -> columns(values.map { it as JsonObject })?.let { Field(name, it) }
        else -> null
    }

    private fun keyedColumns(obj: JsonObject): List<Field>? {
        if (obj.size < 2 || obj.values.any { it !is JsonObject }) return null
        return columns(obj.values.map { it as JsonObject })
    }

    private fun cells(obj: JsonObject, columns: List<Field>): List<String> {
        val out = ArrayList<String>()
        for (column in columns) {
            val value = obj.getValue(column.name)
            if (column.children == null) {
                out += ToonText.primitive(value as JsonPrimitive, delimiter)
            } else {
                out += cells(value as JsonObject, column.children)
            }
        }
        return out
    }

    private fun inline(arr: JsonArray): String =
        arr.joinToString(delimiter.toString()) { ToonText.primitive(it as JsonPrimitive, delimiter) }

    private fun bracket(size: Int): String = "[$size$symbol]"

    private fun emit(depth: Int, text: String) {
        lines += indent(depth) + text
    }

    private fun indent(depth: Int): String = " ".repeat(depth * options.indentSize)
}
